readme.md

```diff
@@ -0,0 +1,183 @@
+> 
+> For UDP sockets, the use of this option can provide better distribution of incoming datagrams to multiple processes (or threads) as compared to the traditional technique of having multiple processes compete to receive datagrams on the same socket.
+> 
+> https://man7.org/linux/man-pages/man7/socket.7.html
+
+![](https://user-images.githubusercontent.com/8400576/127754771-6a789f00-7022-4602-a628-a3d124b43421.png)
+
+As we can notice, we not only obtain the super power to create more than one socket with the same IP: Port combination, we also obtain a kind of load balancer in the kernel mode.
+
+### Go sockets
+
+When we invoke the `net.Listen()` function in Go, this function use the `ListenConfig` struct to create the Listener.
+
+```go
+func Listen(network, address string) (Listener, error) {
+	var lc ListenConfig
+	return lc.Listen(context.Background(), network, address)
+}
+```
+
+If we inspect `ListenConfig,` we can find the method `Control,` and the documentation says: <mark>If Control is not nil, it is called after creating the network connection but before binding it to the operating system.</mark>
+
+```go
+// ListenConfig contains options for listening to an address.
+type ListenConfig struct {
+	// If Control is not nil, it is called after creating the network
+	// connection but before binding it to the operating system.
+	// ...
+	Control func(network, address string, c syscall.RawConn) error
+	...
+}
+```
+
+The function `Control` receives the `syscall.RawConn` which is a raw network connection that has a method also called control (`Control(f func(fd uintptr))`) where it will invoke the function `f` on the underlying connection's file descriptor.
+
+Having the file descriptor we can use the `golang.org/x/sys/unix` package to set the socket options.
+
+```go
+// fd -> the underlying connection's file descriptor.
+// unix.SOL_SOCKET  -> to set options at the socket level, we have to specify the level argument as SOL_SOCKET.
+unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+```
+
+So, we could create a instance of `ListenConfig` with a control function to set the `SO_REUSEPORT` socket options to our sockets.
+
+```go
+var lc = net.ListenConfig{
+	Control: func(network, address string, c syscall.RawConn) error {
+		var opErr error
+		if err := c.Control(func(fd uintptr) {
+			opErr = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+		}); err != nil {
+			return err
+		}
+		return opErr
+	},
+}
+```
+
+### Security
+
+One question we might have at this point is, what about security? I mean, if we can open a socket with the same IP: Port of a specific app, for example, Nginx, we could hijack part of the requests that the kernel will send to us through the socket. Right?v
+
+Well, to prevent this "port hijacking," Linux has special limitations or mechanisms such as:
```

Comment:
Esta oración está rara, usar la palabra "limitations" significa que estás diciendo que Linux "pierde poder", pero no es lo que quieres decir, sino más bien que Linux tiene protecciones o limitaciones para las apps finales... Qué tal si...

```suggestion
Well, to prevent this "port hijacking," Linux has special protections or mechanisms to prevent these problems, such as:
```